flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sinks/TableSinkUtils.scala

```diff
@@ -18,67 +18,99 @@
+    * should be compatible (types should equal including precisions). If types are not compatible,
+    * but can be implicitly casted, a cast projection will be applied. Otherwise, an exception will
+    * be thrown.
+    *
+    * @param query the query to be checked
+    * @param sinkSchema the schema of sink to be checked
+    * @param typeFactory type factory
+    * @return the query RelNode which may be applied the implicitly cast projection.
+    */
+  def validateSchemaAndApplyImplicitCast(
+      query: RelNode,
+      sinkSchema: TableSchema,
+      typeFactory: FlinkTypeFactory,
+      sinkIdentifier: Option[String] = None): RelNode = {
+
+    val queryLogicalType = DataTypeUtils
+      // convert type to nullable, because we ignore nullability when writing query into sink
+      .transform(FlinkTypeFactory.toTableSchema(query.getRowType).toRowDataType, toNullable)
+      .getLogicalType
+      .asInstanceOf[RowType]
+    val sinkLogicalType = DataTypeUtils
+      // convert type to nullable, because we ignore nullability when writing query into sink
+      .transform(sinkSchema.toRowDataType, legacyDecimalToDefaultDecimal, toNullable)
+      .getLogicalType
+      .asInstanceOf[RowType]
+    if (LogicalTypeChecks.areTypesCompatible(queryLogicalType, sinkLogicalType)) {
+      // types are compatible, do nothing
+      query
+    } else if (LogicalTypeCasts.supportsImplicitCast(queryLogicalType, sinkLogicalType)) {
+      // types can be implicit casted, add a cast project
+      val castedDataType = typeFactory.buildRelNodeRowType(
+        sinkLogicalType.getFieldNames,
+        sinkLogicalType.getFields.map(_.getType))
+      RelOptUtils.createCastRel(query, castedDataType)
+    } else {
+      // format query and sink schema strings
+      val srcSchema = queryLogicalType.getFields
+        .map(f => s"${f.getName}: ${f.getType}")
+        .mkString("[", ", ", "]")
+      val sinkSchema = sinkLogicalType.getFields
+        .map(f => s"${f.getName}: ${f.getType}")
+        .mkString("[", ", ", "]")
+
+      val sinkDesc: String = sinkIdentifier.getOrElse("")
+
+      throw new ValidationException(
+        s"Field types of query result and registered TableSink $sinkDesc do not match.\n" +
+          s"Query schema: $srcSchema\n" +
+          s"Sink schema: $sinkSchema")
+    }
+  }
+
+  /**
+    * It checks whether the [[TableSink]] is compatible to the INSERT INTO clause, e.g.
+    * whether the sink is a [[PartitionableTableSink]] and the partitions are valid.
     *
     * @param sinkOperation The sink operation with the query that is supposed to be written.
     * @param sinkIdentifier Tha path of the sink. It is needed just for logging. It does not
     *                      participate in the validation.
     * @param sink     The sink that we want to write to.
     * @param partitionKeys The partition keys of this table.
     */
-  def validateSink(
+  def validateTableSink(
```

Comment:
Can we just add `validateLogicalPhysicalTypesCompatible` to this method and do more check?

Reply:
I think it would better to separate them, because they do the different things and `validateTableSink` is only invoked in "INSERT" statements.